routemaster/models/batch.rb

```diff
@@ -1,56 +1,247 @@
 require 'routemaster/models'
 require 'routemaster/models/message'
+require 'routemaster/models/queue'
+require 'routemaster/models/subscriber'
+require 'routemaster/mixins/redis'
+require 'routemaster/mixins/assert'
 require 'routemaster/mixins/log'
+require 'routemaster/services/codec'
+require 'wisper'
 
 module Routemaster
   module Models
     # Abstracts an ordered list of Message
     class Batch
+      include Mixins::Redis
+      include Mixins::Assert
       include Mixins::Log
-      extend Forwardable
+      include Wisper::Publisher
 
-      def initialize(consumer)
-        @consumer = consumer
-        @batch = []
-        @monitor = Monitor.new
-      end
+      Inconsistency    = Class.new(RuntimeError)
 
-      delegate %i[push length] => :@batch
+      # number of prefix metdata items in a batch list
+      PREFIX_COUNT = 3
 
-      def synchronize(&block)
-        @monitor.synchronize { block.call(self) }
-        self
+
+      attr_reader :uid, :deadline
+
+
+      def initialize(uid:, deadline: nil, subscriber: nil)
+        @uid        = uid
+        @deadline   = deadline
+        @subscriber = subscriber
       end
 
-      def nack
-        synchronize do
-          @batch.each { |msg| @consumer.nack(msg) }
-          _flush
+
+      def subscriber
+        @subscriber ||= begin
+          return if _subscriber_name.nil?
+          Subscriber.find(_subscriber_name)
         end
-        self
       end
 
-      def ack  
-        synchronize do
-          @batch.each { |msg| @consumer.ack(msg) }
-          _flush
+
+      # Return the number of events in the batch (memoised)
+      def length
+        @_length ||= begin
+          raw = _redis.llen(_batch_key)
+          return 0 if raw.nil? || raw.zero?
+          raise Inconsistency, @uid if raw < PREFIX_COUNT
+          raw - PREFIX_COUNT
         end
+      end
+
+
+      # Does this batch still exist?
+      def exists?
+        _redis.exists(_batch_key)
+      end
+
+
+      # Is this batch deliverable?
+      def valid?
+        subscriber && data&.any?
```

Comment:
Ruby 2.1 introduced the `&.` operator; this is conviently equivalent to (and slightly faster than)

```ruby
subscriber && (data && data.any?)
```